File: problems/06_linked_list/0019_remove_nth_node_from_end_of_list/solution.swift
// Problem: 0019. Remove Nth Node From End of List
// Difficulty: Medium

import Foundation

class ListNode {
    var val: Int; var next: ListNode?
    init(_ v: Int = 0, _ n: ListNode? = nil) { val = v; next = n }
}

func fromArr(_ a: [Int]) -> ListNode? {
    let d = ListNode(); var c = d
    for v in a { c.next = ListNode(v); c = c.next! }
    return d.next
}

func toArr(_ h: ListNode?) -> [Int] {
    var out: [Int] = []; var p = h
    while p != nil { out.append(p!.val); p = p!.next }
    return out
}

class Solution {
    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        let dummy = ListNode(0, head)
        var fast: ListNode? = dummy, slow: ListNode? = dummy
        for _ in 0..<(n + 1) { fast = fast!.next }
        while fast != nil { fast = fast!.next; slow = slow!.next }
        slow!.next = slow!.next!.next
        return dummy.next
    }
}

func checkArr(_ a: [Int], _ e: [Int], _ msg: String) {
    if a != e { fatalError("FAIL \(msg)") }
}

let sol = Solution()
checkArr(toArr(sol.removeNthFromEnd(fromArr([1, 2, 3, 4, 5]), 2)), [1, 2, 3, 5], "ex1")
checkArr(toArr(sol.removeNthFromEnd(fromArr([1]), 1)), [], "one")
checkArr(toArr(sol.removeNthFromEnd(fromArr([1, 2]), 1)), [1], "tail")
checkArr(toArr(sol.removeNthFromEnd(fromArr([1, 2]), 2)), [2], "head")
checkArr(toArr(sol.removeNthFromEnd(fromArr([1, 2, 3, 4, 5]), 5)), [2, 3, 4, 5], "head5")
checkArr(toArr(sol.removeNthFromEnd(fromArr([1, 2, 3, 4, 5]), 1)), [1, 2, 3, 4], "tail5")
print("All tests passed!")

File: problems/06_linked_list/0019_remove_nth_node_from_end_of_list/solution.py
"""
Problem: 0019. Remove Nth Node From End of List
Difficulty: Medium
URL: https://leetcode.com/problems/remove-nth-node-from-end-of-list/

解法：dummy + 快慢指標一次掃描，O(L)/O(1)。
"""
from __future__ import annotations
from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None):
        self.val = val
        self.next = next


def to_list(h):
    out = []
    while h: out.append(h.val); h = h.next
    return out


def from_list(arr):
    d = ListNode(); c = d
    for v in arr: c.next = ListNode(v); c = c.next
    return d.next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        fast = slow = dummy
        for _ in range(n + 1):
            fast = fast.next
        while fast:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return dummy.next

    def removeNthFromEnd_brute(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        L = 0
        cur = head
        while cur:
            L += 1; cur = cur.next
        target = L - n  # 0-indexed
        dummy = ListNode(0, head)
        prev = dummy
        for _ in range(target):
            prev = prev.next
        prev.next = prev.next.next
        return dummy.next


if __name__ == "__main__":
    sol = Solution()
    for fn in (sol.removeNthFromEnd, sol.removeNthFromEnd_brute):
        assert to_list(fn(from_list([1, 2, 3, 4, 5]), 2)) == [1, 2, 3, 5]
        assert to_list(fn(from_list([1]), 1)) == []
        assert to_list(fn(from_list([1, 2]), 1)) == [1]
        assert to_list(fn(from_list([1, 2]), 2)) == [2]  # 刪頭
        assert to_list(fn(from_list([1, 2, 3, 4, 5]), 5)) == [2, 3, 4, 5]  # 刪頭
        assert to_list(fn(from_list([1, 2, 3, 4, 5]), 1)) == [1, 2, 3, 4]  # 刪尾
    print("All tests passed!")
